site/content/integrate/apps/functionality/oauth2.md

```diff
@@ -0,0 +1,269 @@
+func oauth2Complete(w http.ResponseWriter, req *http.Request) {
+	// Decode the call request data
+	callRequest := new(apps.CallRequest)
+	err := json.NewDecoder(req.Body).Decode(callRequest)
+	if err != nil {
+		// handle the error
+	}
+	// Read the authorization code from the call request values
+	authCodeIntf, ok := callRequest.Values["code"]
+	if !ok {
+		// handle the error
+	}
+	authCode, ok := authCodeIntf.(string)
+	if !ok {
+		// handle the error
+	}
+	// Exchange the authorization code for a token
+	token, err := oauth2Config(callRequest).Exchange(context.Background(), authCode)
+	if err != nil {
+		// handle the error
+	}
+	// Get an instance of the API client as the acting user
+	client := appclient.AsActingUser(callRequest.Context)
+	// Store the OAuth2 token for the acting user
+	err = client.StoreOAuth2User(token)
+	if err != nil {
+		// handle the error
+	}
+	// Return a success response
+	err = httputils.WriteJSON(w, apps.NewDataResponse(nil))
+	if err != nil {
+		// handle the error
+	}
+}
+```
+
+#### OAuth2 configuration
+
+The Golang `oauth2Config` function referenced in the example above is responsible for creating a `oauth2.Config` ({{<newtabref title="godoc" href="https://pkg.go.dev/golang.org/x/oauth2#Config">}}) struct which contains the necessary information to exchange an authorization code for a token.
+
+Example Golang code for the `oauth2Config` function referencing a Google OAuth2 provider:
+
+```go
+func oauth2Config(creq *apps.CallRequest) *oauth2.Config {
+    return &oauth2.Config{
+        ClientID:     creq.Context.OAuth2.ClientID,
+        ClientSecret: creq.Context.OAuth2.ClientSecret,
+        Endpoint:     google.Endpoint,
+        RedirectURL:  creq.Context.OAuth2.CompleteURL,
+        Scopes: []string{
+            "https://www.googleapis.com/auth/calendar",
+            "https://www.googleapis.com/auth/userinfo.profile",
+            "https://www.googleapis.com/auth/userinfo.email",
+        },
+    }
+}
+```
+
+### Get a user access token with a driver
+
+Example [Golang driver]({{<ref "/integrate/apps/drivers/golang">}}) code to get a user access token:
+
+```go
+func useRemoteService(w http.ResponseWriter, req *http.Request) {
+	// Decode the call request data
+	callRequest := new(apps.CallRequest)
+	err := json.NewDecoder(req.Body).Decode(callRequest)
+	if err != nil {
+		// handle the error
+	}
+	// Get an instance of the API client as the acting user
+	client := appclient.AsActingUser(callRequest.Context)
+	// Get the user's token
+	var userToken string
+	err = client.GetOAuth2User(&userToken)
+	if err != nil {
+		// handle the error
+	}
+	// Use the token
+	// ...
+	// Return a success response
+	err = httputils.WriteJSON(w, apps.NewDataResponse(nil))
+	if err != nil {
+		// handle the error
+	}
+}
+```
+
+{{<note>}}
+The user's access token can also be obtained from the call request context `oauth2_user` field. The call must be configured to expand the `oauth2_user` context field.
```

Comment:
https://github.com/mattermost/mattermost-plugin-apps/pull/397

```suggestion
The user's access token can also be obtained from the call request context `oauth2_user` field. The call must be configured to expand the `oauth2_user` context field with expand level `all`.
```

Reply:
@mickmister What does the `summary` level give? Should it not also give this information? Is there a `summary` level for this expand field at all?

Reply:
Currently there is nothing provided for `summary`. From https://github.com/mattermost/mattermost-plugin-apps/pull/397#issuecomment-1323555279:

> Coming back to this change: 2/5 that I like the current behaviour as it's'more strict and allows us to add a meaningful use of `summary` later.

Reply:
Why does this code include the `summary` level in its check, then?
https://github.com/mattermost/mattermost-plugin-apps/blob/0856d8f8eac70040d6715f7c3733a3d3ca288ddc/server/proxy/expand.go#L138

Reply:
Please see the code on the master branch https://github.com/mattermost/mattermost-plugin-apps/blob/2e3c62c9a83a2a5c899576bd5db78922164acd14/server/proxy/expand.go#L141

Reply:
Any discussion about this particular issue can be discussed here where there is already discussion captured https://github.com/mattermost/mattermost-plugin-apps/pull/397

Reply:
We can only document released functionality in this PR.

https://github.com/mattermost/mattermost-plugin-apps/pull/397 has the label "Docs/Not Needed". Should it have the "Docs/Needed" label instead?

Reply:
The reason it has "Docs/Not Needed" is because the PR was closed. The "Docs/Needed" label is on the PR that introduced the changes however https://github.com/mattermost/mattermost-plugin-apps/pull/336

Reply:
> We can only document released functionality in this PR.

I don't want to have outdated information that doesn't work though. We should assume 1.2 is released since we are releasing it very soon to resolve issues like this

Reply:
@levb @hanzei Thoughts on including documentation that relates to v1.2 of the framework in this PR?

Reply:
Our developer documentation should detail what's released and available publicly and should not include future functionality not yet released. Doing so can damage trust we're building up among our developer and integration communities. 

Please hold off on releasing 1.2 documentation until 1.2 has been released. CC @asatkinson

Reply:
@cwarnermm In this case, there is a breaking change to something not fully fleshed out previously. This specific issue has significantly affected partnering orgs creating Apps and our internally developed Apps because of this detail. It's simply a "bug fix" I'm trying to put into the documentation here.

I understand the point of making documentation consistent with what's available, but 5/5 it's a moot point in this exact situation. I proposed that this breaking change may constitute a major version of the App framework, but the team decided not to so this, since the feature had not been flshed out yet https://github.com/mattermost/mattermost-plugin-apps/pull/397#issuecomment-1323555279. Please see the discussion on that PR for more details